### PR TITLE
feat(agents): add env var overrides for agent configs

### PR DESCRIPTION
**Pull request by lattejed**
Agent config should be improved to speed up deployments. Currently most config changes have to go through a PR->deploy cycle. Configs should be updatable after deployment

## Solution

Agent config should be overridable with env vars

## PR Checklist

- [x] Added Tests
- [x] Updated Documentation N/A
- [x] Updated CHANGELOG.md for the appropriate package

### CODE REVIEW

**Comment by lattejed** [comment]
@ltchang2019 I'm pretty sure there are no in-repo docs that have to get updated for this. If I missed anything lmk.

File: fixtures/env.test-agents
Comment (ltchang2019) [inline fixtures/env.test-agents:7]: @arnaud036 thoughts on using comma separated string for loading list through env vars? I know we currently use `REPLICA_{number}_NAME` for the replica list but it gets a bit crazy for agent config lists. Thoughts on?
- Switching loading replicas to comma-separated list + using comma list for agent configs
- Using indexed entries like replica for agent configs
- Using indexed entries for replica list BUT comma separated list of agent configs

File: configuration/src/agent/mod.rs
Comment (ltchang2019) [inline configuration/src/agent/mod.rs:80]: Feels a bit awkward to implement this as no-op function for [$name Config]. Thinking that declaration of env overrides should happen in the `decl_config!` macro rather than in a separate one, as we want every agent config to be overridable.

Comment (ltchang2019) [reply configuration/src/agent/mod.rs:80]: Think the best way to try going about this would be to have every type within an agent config struct implement something like `FromEnv`. Then, in `decl_config!`, we could iterate through all the struct in fields, look for their corresponding env vars using `from_env()`, then load them and modify `&mut self`. Note that for fields that are `HashSet<T>` or `HashMap<T>`, we can implement custom util functions that call `<T>::from_env()` in a loop.

This way, the custom work lies within loading the struct fields themselves but loading the whole struct from env is DRYed up using macro.

Comment (ltchang2019) [reply configuration/src/agent/mod.rs:80]: Something like:
```
macro_rules! decl_config {
    (
        $name:ident {
            $($(#[$tags:meta])* $prop:ident: $type:ty,)*
        }
    ) => {
        ... existing [$name Config] struct declaration ...

        impl [$name Config] {
	fn load_env_overrides(&mut self) {
          $(
              {
                 let field = <$prop>::from_env();
                 self.$prop = field;
              }
           )*
	}
    }
}
```

**Review by ltchang2019 — COMMENTED** [review]
Might want to try slightly different approach as noted in comment below. Approach would basically be to have agent config struct _fields_ implement `FromEnv` and then use macro to DRY up code that goes through each struct field, loads it from env, then overrides `&mut self`.

**Comment by ltchang2019** [comment]
Also we're not calling `load_env_overrides` anywhere right now. We want to load env overrides on agent config in `nomad-base/src/settings/macros.rs` below [this line](https://github.com/nomad-xyz/rust/blob/main/nomad-base/src/settings/macros.rs#L107). That way everytime agent loads up settings, it tries to override using env vars.

**Comment by lattejed** [comment]
> Also we're not calling `load_env_overrides` anywhere right now. We want to load env overrides on agent config in `nomad-base/src/settings/macros.rs` below [this line](https://github.com/nomad-xyz/rust/blob/main/nomad-base/src/settings/macros.rs#L107). That way everytime agent loads up settings, it tries to override using env vars.

It's called here: 

https://github.com/nomad-xyz/rust/blob/115b4b56ac11858124a827fb7df4b07b3020601d/nomad-base/src/settings/macros.rs#L110

The passing integration test is here:

https://github.com/nomad-xyz/rust/blob/cc0e3852f389f13f160b45ab0cd38fd28ed3dee8/agents/kathy/src/settings.rs#L21

File: configuration/src/agent/mod.rs
Comment (lattejed) [reply configuration/src/agent/mod.rs:80]: @ltchang2019 We're deserializing here so the only _correct_ way to do this is to extend `serde` -- then we can have arbitrarily complex configs and it will just work.

The way to do that would be to extend e.g., `envy` -- use it as a jumping off point, add it to an existing crate so we don't have the overhead of another (internal) dep. It's a pretty thin impl anyway. And then extend it to our use cases.

As I suggested elsewhere, this is at best an 'ok' solution. But it's correct and tested and (minus tests) it's ~100 LOC and easy enough to get rid of when we have time to do the above. 

Given that we need to solidify this interface let's use this (once we get input from @arnaud036) and then do it properly when we have more time for it.

Comment (lattejed) [reply configuration/src/agent/mod.rs:80]: As discussed, this needs an RFC about where it should land when we refactor it. I'm on that

File: fixtures/env.test-agents
Comment (arnaud036) [inline fixtures/env.test-agents:13]: Same confusion as for KATHY_ENABLED

File: fixtures/env.test-agents
Comment (lattejed) [reply fixtures/env.test-agents:11]: The approach here was to expose the entirety of the config struct. Are we 100% no one (internal or external) will ever want to override this? Also, I'm not sure if it makes sense to trim it since it's already in there.

File: fixtures/env.test-agents
Comment (lattejed) [reply fixtures/env.test-agents:5]: All agents have `*_INTERVAL` and `*_ENABLED`. As above: The approach here was to expose the entirety of the config struct.

File: fixtures/env.test-agents
Comment (arnaud036) [reply fixtures/env.test-agents:11]: Ok sounds good. Since, it's not required, it makes sense to keep this feature around

File: configuration/src/agent/mod.rs
Comment (ltchang2019) [inline configuration/src/agent/mod.rs:58]: Can't this just be `&mut self` not `$self_`?

**Review by ltchang2019 — COMMENTED** [review]
Couple things before merging :)
1. Rebase on main and update necessary changelogs (we have changelogs for all crates now so agents need changes)
2. Link to [article on identifiers](https://veykril.github.io/tlborm/decl-macros/minutiae/identifiers.html) and `self` keyword above `decl_env_overrides`
3. Run in dev as per usual

**Review by ltchang2019 — APPROVED** [review]
Approving with understanding that this is temporary solution to be eventually replaced by longer-term env var deserialization scheme

**Pull request merged**
